src/main/tut/simple-algebraic-data-types.md

```diff
@@ -0,0 +1,366 @@
+```
+...............
+```Haskell
+startsWithSymbol :: (String, String, Int) -> Bool
+startsWithSymbol (name, symbol, _) = isPrefixOf symbol name
+```
+```tut:silent
+val startsWithSymbol: (String, String, Int) => Boolean =
+  (name, symbol, _) => name.startsWith(symbol)
+```
+...............
+```Haskell
+data Element = Element { name         :: String
+                       , symbol       :: String
+                       , atomicNumber :: Int }
+```
+```tut:silent
+case class Element(
+  name: String,
+  symbol: String,
+  atomicNumber: Int
+)
+```
+...............
+```Haskell
+tupleToElem :: (String, String, Int) -> Element
+tupleToElem (n, s, a) = Element { name = n
+                                , symbol = s
+                                , atomicNumber = a }
+```
+```tut:silent
+val tupleToElem: (String, String, Int) => Element =
+  (n, s, a) => Element(n, s, a)
+```
+...............
+```Haskell
+elemToTuple :: Element -> (String, String, Int)
+elemToTuple e = (name e, symbol e, atomicNumber e)
+```
+```tut:silent
+val elemToTuple: Element => (String, String, Int) =
+  e => (e.name, e.symbol, e.atomicNumber)
+```
+...............
+```Haskell
+atomicNumber :: Element -> Int
+```
+```scala
+val atomicNumber: Element => Int
+```
+...............
+```Haskell
+startsWithSymbol :: Element -> Bool
+startsWithSymbol e = isPrefixOf (symbol e) (name e)
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+startsWithSymbol e = symbol e `isPrefixOf` name e
+```
+```tut:silent
+val startsWithSymbol: Element => Boolean =
+  e => e.name startsWith e.symbol
+```
+...............
+```Haskell
+data Either a b = Left a | Right b
+```
+```tut:silent
+sealed trait Either[+A, +B]
+case class Left[+A](a: A) extends Either[A, Nothing]
+case class Right[+B](b: B) extends Either[Nothing, B]
+```
+...............
+```Haskell
+data OneOfThree a b c = Sinistral a | Medial b | Dextral c
+```
+```tut:silent
+sealed trait OneOfThree[A, B, C]
+case class Sinistral[A](a: A) extends OneOfThree[A, Nothing, Nothing]
+case class Medial[B](b: B) extends OneOfThree[Nothing, B, Nothing]
+case class Dextral[C](c: C) extends OneOfThree[Nothing, Nothing, C]
+```
+...............
+```Haskell
+Either a Void
+```
+```scala
+Either[A, Nothing]
+```
+...............
+```Haskell
+data Color = Red | Green | Blue
+```
+```tut:silent
+sealed trait Color
+case object Red extends Color
+case object Green extends Color
+case object Blue extends Color
+```
+...............
+```Haskell
+data Bool = True | False
+```
+```tut:silent
+sealed trait Bool
+case object True extends Bool
+case object False extends Bool
+```
+...............
+```Haskell
+data Maybe a = Nothing | Just a
+```
+```tut:silent
+sealed trait Maybe[+A]
+case object Nothin extends Maybe[Nothing]
+case class Just[+A](a: A) extends Maybe[A]
+```
+...............
+```Haskell
+data NothingType = Nothing
+```
+```tut:silent
+trait NothingType[Nothing]
```

Comment:
This one is tricky. `Nothing` here becomes a type parameter name, not sure if that's what you want. 
Should it be 
```
sealed trait NothingType
object Nothing extends NothingType
```

Reply:
My comment got screwed up by github 😞 
Anyway I agree with @kailuowang on this one, except we should call it `NoneType`. It also may be worth adding a note stating that the distinction between type and data constructors is not strong in Scala. The code above is arguably unidiomatic, who actually writes a sealed trait with one case only?

Reply:
@kailuowang Ah, good call - forgot about the constructor name. Also the fact that Haskell's `Nothing` =/= Scala's `Nothing` 😅

@SystemFw I agree - the C++ examples tend to have little comments here and there as well to help explain what's going on, so I'm sure Bartosz wouldn't mind us doing a bit of the same with these sorta awkward bits.